### PR TITLE
FISH-9513 obfuscate password in domainxml

### PR DESCRIPTION
**Pull request by NotedSalmon**
I have made changes to the domainXmlCollector so it parses the xml file and checks for a node with password and changes the value of it and it also checks the name nodes to see if they include "Password" and changes the value of that to PASSWORD_HIDDEN.

I have covered the following use cases using the `.contains()`:
```
        Line   4:     <authentication-service default="true" name="adminAuth" use-password-credential="true">
	Line  52:       <property name="Password" value="password_placeholder"></property>
	Line  54:     <mail-resource password="password1" auth="true" host="test@test.com" description="This is a description" from="test@test.com" user="Hello" jndi-name="test"></mail-resource>
	Line 148:         <property name="search-bind-password" value="plain-text-password"></property>
	Line 191:             <property name="dynamic.username.password" value="false"></property>
	Line 199:             <property name="dynamic.username.password" value="false"></property>
	Line 411:             <property name="dynamic.username.password" value="false"></property>
	Line 419:             <property name="dynamic.username.password" value="false"></property>
	Line 600:   <property name="payara.microprofile.admin.password.12" value="passwoeoeoased"></property>
```
After:
```
        Line    8:         <authentication-service default="true" name="adminAuth" use-password-credential="true">
	Line  104:             <property name="Password" value="PASSWORD_HIDDEN"/>
	Line  108:         <mail-resource auth="true" description="This is a description" from="test@test.com" host="test@test.com" jndi-name="test" password="PASSWORD_HIDDEN" user="Hello"/>
	Line  296:                 <property name="search-bind-password" value="PASSWORD_HIDDEN"/>
	Line  382:                         <property name="dynamic.username.password" value="PASSWORD_HIDDEN"/>
	Line  398:                         <property name="dynamic.username.password" value="PASSWORD_HIDDEN"/>
	Line  822:                         <property name="dynamic.username.password" value="PASSWORD_HIDDEN"/>
	Line  838:                         <property name="dynamic.username.password" value="PASSWORD_HIDDEN"/>
	Line 1200:     <property name="payara.microprofile.admin.password.12" value="PASSWORD_HIDDEN"/>
```

### CODE REVIEW

File: src/main/java/fish/payara/extras/diagnostics/collection/collectors/DomainXmlCollector.java [hunk 1]
```diff
@@ -1,45 +1,6 @@
...
- * elects to include this software in this distribution under the [CDDL or GPL
- * Version 2] license."  If you don't indicate a single choice of license, a
- * recipient has the option to distribute your version of this file under
- * either the CDDL, the GPL Version 2 or to extend the choice of license to
- * its licensees as provided above.  However, if you add GPL Version 2 code
- * and therefore, elected the GPL Version 2 license, then the option applies
- * only if the new code is made subject to such option by the copyright
- * holder.
- */
-
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/collection/collectors/DomainXmlCollector.java:40]: Keep the copyright

File: src/main/java/fish/payara/extras/diagnostics/collection/collectors/FileCollector.java [hunk 2]
```diff
@@ -82,21 +82,25 @@ public FileCollector(File file, String destination) {
     public int collect() {
         try {
             if (confirmPath(filePath, false) && confirmPath(destination, true)) {
-                if (instanceName != null) {
-                    Files.copy(filePath, destination.resolve((instanceName + "-" + filePath.getFileName())), REPLACE_EXISTING);
-                } else {
-                    Files.copy(filePath, destination.resolve(filePath.getFileName()), REPLACE_EXISTING);
-                }
+                Path targetFile = resolveDestinationFile();
+                Files.copy(filePath,targetFile, REPLACE_EXISTING);
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/collection/collectors/FileCollector.java:86]:
```suggestion
                Files.copy(filePath, targetFile, REPLACE_EXISTING);
```

File: src/main/java/fish/payara/extras/diagnostics/util/ObfuscateDomainXml.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import java.io.File;
+import java.util.logging.Logger;
+
+public class ObfuscateDomainXml {
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/util/ObfuscateDomainXml.java:18]: Copyright and the name could be better - something like 'DomainXmlUtil'

**Review by kalinchan — CHANGES_REQUESTED** [review]

File: src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) 2023-2024 Payara Foundation and/or its affiliates. All rights reserved.
```
Comment (jGauravGupta) [inline src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java:4]: Please fix the copyright year.

**Review by jGauravGupta — CHANGES_REQUESTED** [review]
````
<jms-service type="EMBEDDED" default-jms-host="default_JMS_host" addresslist-behavior="priority">
      <jms-host name="default_JMS_host" host="localhost" port="${JMS_PROVIDER_PORT}" admin-user-name="admin" admin-password="admin" lazy-init="true" />
</jms-service>
 ````
 
 In this example, password is stored via `admin-password` property.
 
 
 ````
 <auth-realm name="ldap-realm" classname="com.sun.enterprise.security.auth.realm.ldap.LDAPRealm">
    <property name="directory" value="ldap://localhost:389" />
    <property name="base-dn" value="dc=example,dc=com" />
    <property name="search-filter" value="(uid=%s)" />
    <property name="search-bind-dn" value="cn=Directory Manager" />
    <property name="search-bind-password" value="plain-text-password" />
</auth-realm>
````


 In this example, password is stored via `search-bind-password` property.

File: src/main/java/fish/payara/extras/diagnostics/asadmin/CollectAsadmin.java [hunk 5]
```diff
@@ -70,6 +70,9 @@ public class CollectAsadmin extends BaseAsadmin {
     @Param(name = ParamConstants.DOMAIN_XML_PARAM, optional = true, defaultValue = "true")
     private boolean collectDomainXml;
 
+    @Param(name = ParamConstants.OBFUSCATE_DOMAIN_XML_PARAM, optional = true, defaultValue = "true")
```
Comment (jGauravGupta) [inline src/main/java/fish/payara/extras/diagnostics/asadmin/CollectAsadmin.java:73]: As per ticket description "An option should be added to the collect-diagnostics command: --`obfuscate` true|false " but the parameter name is currently `obfuscateDomainXml`. Please update the parameter name to match the description for consistency.

**Review by jGauravGupta — CHANGES_REQUESTED** [review]

File: src/main/java/fish/payara/extras/diagnostics/asadmin/CollectAsadmin.java [hunk 6]
```diff
@@ -70,6 +70,9 @@ public class CollectAsadmin extends BaseAsadmin {
...
+    private boolean obfuscateDomainXml;
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/asadmin/CollectAsadmin.java:74]: Might be better to name this something more generic.

Perhaps just `obfuscate`

File: src/main/java/fish/payara/extras/diagnostics/collection/collectors/DomainXmlCollector.java [hunk 7]
```diff
@@ -52,32 +53,39 @@ public class DomainXmlCollector extends FileCollector {
     private Path path;
     private String dirSuffix;
     private Logger LOGGER = Logger.getLogger(DomainXmlCollector.class.getName());
+    private boolean obfuscateDomainXml;
+    private final int COLLECTED_OKAY = 0;
+    private final DomainXmlUtil obfuscate = new DomainXmlUtil();
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/collection/collectors/DomainXmlCollector.java:58]:
```suggestion
    private final DomainXmlUtil domainXmlUtil = new DomainXmlUtil();
```

File: src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+public class DomainXmlUtil {
+    private static final Logger LOGGER = Logger.getLogger(DomainXmlUtil.class.getName());
+    private static final String PASSWORD_CHANGE = "PASSWORD_HIDDEN";
+    private static final String PASSWORD_KEYWORD = "password";
+    private static final String ADMIN_PASSWORD_KEYWORD = "admin-password";
+    private static final String SEARCH_BIND_KEYWORD = "search-bind-password";
+    private static final String NAME_KEYWORD = "name";
+    private static final String VALUE_KEYWORD = "value";
+
+    public void obfuscateDomainXml(File xmlFile) {
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java:67]:
```suggestion
    public void obfuscateDomainXml (File xmlFile) {
```

File: src/main/java/fish/payara/extras/diagnostics/util/ParamConstants.java [hunk 9]
```diff
@@ -65,6 +65,7 @@ public final class ParamConstants {
     //Option Params
     public static final String DOMAIN_NAME = "DomainName";
     public static final String DOMAIN_XML_FILE_PATH = "DomainXMLFilePath";
+    public static final String OBFUSCATE_DOMAIN_XML_PARAM = "obfuscate";
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/util/ParamConstants.java:68]:
```suggestion
    public static final String OBFUSCATE_PARAM = "obfuscate";
```

**Review by kalinchan — CHANGES_REQUESTED** [review]

File: src/main/java/fish/payara/extras/diagnostics/util/ParamConstants.java (hunk 9 above)
Comment (kalinchan) [reply src/main/java/fish/payara/extras/diagnostics/util/ParamConstants.java:68]: Also move this to the param section

File: src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            if (hasPasswordAttribute) {
+                tempNode.setAttribute(PASSWORD_KEYWORD, PASSWORD_CHANGE);
+            }
+            if (hasAdminPasswordAttribute) {
+                tempNode.setAttribute(ADMIN_PASSWORD_KEYWORD, PASSWORD_CHANGE);
+            }
+
+            String nameAttribute = tempNode.getAttribute(NAME_KEYWORD);
+            boolean hasValueAttribute = tempNode.hasAttribute(VALUE_KEYWORD);
+            if (PASSWORD_KEYWORD.equalsIgnoreCase(nameAttribute) || SEARCH_BIND_KEYWORD.equalsIgnoreCase(nameAttribute)) {
```
Comment (kalinchan) [inline src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java:106]: Is this not missing ADMIN_PASSWORD_KEYWORD?

Comment (NotedSalmon) [reply src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java:106]: No this is for passwords that are within properties so like name="Password" value="password__"

Comment (jGauravGupta) [reply src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java:106]: Use regex to cover multiple scenarios like custom properties in a custom auth config, example:
````
        String regex = "[^\"]*password[^\"]*";
        Pattern pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(text);
        if (matcher.find()) {
            System.out.println(matcher.group());
        }
````

**Review by jGauravGupta — CHANGES_REQUESTED** [review]

File: src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java (hunk 10 above)
Comment (NotedSalmon) [reply src/main/java/fish/payara/extras/diagnostics/util/DomainXmlUtil.java:106]: I used it for the name using the `.contains()`. I have added the scenarios to the description.

**Review by jGauravGupta — APPROVED** [review]

**Review by kalinchan — APPROVED** [review]

**Pull request merged**